gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -0,0 +1,281 @@
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.source.extractor.extract.kafka;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.SourceState;
+import gobblin.configuration.WorkUnitState;
+import gobblin.source.extractor.Extractor;
+import gobblin.source.extractor.extract.EventBasedSource;
+import gobblin.source.workunit.Extract;
+import gobblin.source.workunit.WorkUnit;
+
+import com.google.common.base.Splitter;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.io.Closer;
+
+
+/**
+ * Kafka source implementation
+ *
+ * @author ziliu
+ *
+ */
+public class KafkaSource extends EventBasedSource<Schema, GenericRecord> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaSource.class);
+
+  public static final String TOPIC_BLACKLIST = "topic.blacklist";
+  public static final String TOPIC_WHITELIST = "topic.whitelist";
+  public static final String TOPICS_MOVE_TO_LATEST_OFFSET = "topics.move.to.latest.offset";
+  public static final String MOVE_TO_EARLIEST_OFFSET_ALLOWED = "move.to.earliest.offset.allowed";
+  public static final boolean DEFAULT_MOVE_TO_EARLIEST_OFFSET_ALLOWED = false;
+  public static final String TOPIC_NAME = "topic.name";
+  public static final String PARTITION_ID = "partition.id";
+  public static final String LEADER_ID = "leader.id";
+  public static final String LEADER_HOST = "leader.host";
+  public static final String LEADER_PORT = "leader.port";
+  public static final Extract.TableType DEFAULT_TABLE_TYPE = Extract.TableType.APPEND_ONLY;
+  public static final String DEFAULT_NAMESPACE_NAME = "KAFKA";
+
+  private final Set<String> moveToLatestTopics = Sets.newTreeSet(String.CASE_INSENSITIVE_ORDER);
+  private final Map<KafkaPartition, Long> previousOffsets = Maps.newHashMap();
+
+  @Override
+  public List<WorkUnit> getWorkunits(SourceState state) {
+    List<WorkUnit> workUnits = Lists.newArrayList();
+    Closer closer = Closer.create();
+    try {
+      KafkaWrapper kafkaWrapper = closer.register(KafkaWrapper.create(state));
+      List<KafkaTopic> topics = getFilteredTopics(kafkaWrapper, state);
+      for (KafkaTopic topic : topics) {
+        workUnits.addAll(getWorkUnitsForTopic(kafkaWrapper, topic, state));
+      }
+      return workUnits;
+    } finally {
+      try {
+        closer.close();
```

Comment:
`closer.close()` throws IOException, but `Source.getWorkunits()` doesn't. So need to wrap it here.